### PR TITLE
:sparkles: Log in Tailscale users automatically on staging servers

### PR DESCRIPTION
**Pull request by Marigold**
## Motivation

We log in as admin user on staging servers. This is a bit annoying because for [data managers](https://github.com/owid/etl/issues/2578) because:
- Each time we land on a new staging server, we have to log in with admin (only once per staging server) and then we see a warning about "compromised passwords" on Chrome
- Chart edits on staging server are all made by "ETL" user (this is problematic since a lot of chart edits come from staging servers via chart-sync)

## Implementation

Add middleware that compares user IP to IPs from `tailscale status --json`. If we find a match, we'll look up Tailscale display name (which is actually a **Github name**) and use it to look up that user in MySQL. This assumes that Github names match MySQL full names (which is true for most people, and exceptions can change it).

Once we identify user, we call `logInAsUser` and set `sessionid` in cookies. `authMiddleware` then uses the sessionid and logs us automatically.

## How to test

1. Open admin at http://staging-site-real-users/admin in **incognito mode**
2. You should be automatically logged in

## How to develop

1. Go to `ssh owid@staging-site-real-users` in terminal or VSCode
3. `pm2 stop admin`, then `cd owid-grapher && yarn startAdminDevServer`
4. Open admin at http://staging-site-real-users/admin in anonymous mode
5. You should be logged automatically and shouldn't see a login page

## Github names not matching `users.fullName`

Tuna -> Tuna Acisu
bastianherre -> Bastian Herre
JoeHasell -> Joe Hasell
mrwbkrm -> Marwa Boukarim
veronikasamborska1994 -> Veronika Samborska

### CODE REVIEW

**Comment by rakyi** [comment]
Have you verified that the session exists when read in the `authMiddleware`?

**Comment by rakyi** [comment]
In general, I think it would be better to use the same way to log in into staging that we use for production. Is that not possible?

**Comment by Marigold** [comment]
> In general, I think it would be better to use the same way to log in into staging that we use for production. Is that not possible?

I haven't tried it, but my guess is that it'd be generally more complex. We'd need to set up Cloudflare tunnels & auths for staging servers (possibly dynamically through API) and start accessing servers via https://staging-site-branch.owid.io or something instead of http://staging-site. I was hoping that grabbing Tailscale user would make it all unnecessary.

**Comment by Marigold** [comment]
@marcelgerber, could you please take a look and test it? (See 'how to test' above.) You seem to be the most knowledgeable about our auth and everything in general. ;)

**Comment by rakyi** [comment]
Sorry for the slow response, I'm prioritizing my own work currently. Hopefully Marcel is less busy. At worse, I should be able to get back to this next week.

File: adminSiteServer/authentication.tsx [hunk 1]
```diff
@@ -250,3 +251,125 @@ export async function logInWithCredentials(
...
+    }
+    return ip
+}
+
+async function getTailscaleIpToUserMap(): Promise<Record<string, string>> {
+    return new Promise((resolve) => {
+        exec("tailscale status --json", (error, stdout) => {
+            if (error) {
+                console.error(`Error getting Tailscale status: ${error}`)
+                return resolve({})
```
Comment (marcelgerber) [inline adminSiteServer/authentication.tsx:341]: I think it's better if this `reject`s in an error, and the calling site needs to make sure it catches this (probably just by using `.catch(() => ({}))`).

**Review by marcelgerber — APPROVED** [review]
Nice, this works well!

I was wondering about two things while looking at this:
- Why have you opted to not use the seemingly simpler [Tailscale request headers](https://tailscale.com/kb/1312/serve#identity-headers), like `Tailscale-User-Login`?
  I guess there is a risk that they could be faked if the incoming request is not coming from Tailscale - but potentially we could check for that some way and/or make sure that this can never happen?
- Do you think it would make sense to include a new column in our users table, something like `githubUsername` (or, alternatively, `tailscaleUsername`). That seems a bit less brittle than the display name mapping we're doing here?

**Pull request merged**